jdaviz/core/template_mixin.py

```diff
@@ -68,3 +75,219 @@ def _mxn_update_plugin_opened(self, new_value):
+                break
+        else:
+            color = False
+        return {"label": subset.label, "color": color}
+
+    def _delete_spectral_subset(self, msg):
+        # NOTE: calling .remove will not trigger traitlet update
+        self.items = [s for s in self.items
+                      if s['label'] != msg.subset.label]
+        if self.selected not in self.labels:
+            self.selected = "Entire Spectrum"
+
+    def _update_spectral_subset(self, msg):
+        if isinstance(msg.subset.subset_state, RoiSubsetState):
+            # then this is a spatial subset, we want to ignore
+            return
+
+        if msg.subset.label not in self.labels:
+            # NOTE: += will not trigger traitlet update
+            self.items = self.items + [self._subset_to_dict(msg.subset)]  # noqa
+        else:
+            if msg.attribute in ('style'):
+                # TODO: may need to add label and then rebuild the entire list if/when
+                # we add support for renaming subsets
+
+                # NOTE: in-line replacement (self.spectral_subset_items[i] = ...)
+                # will not trigger traitlet update
+                self.items = [s if s['label'] != msg.subset.label
+                              else self._subset_to_dict(msg.subset)
+                              for s in self.items]
+
+        if msg.attribute == 'subset_state' and msg.subset.label == self.selected:
+            # updated the currently selected subset
+            self._clear_cache("selected_obj")
+            self._update_has_subregions()
+
+    def _selected_changed(self, event):
+        if event['new'] not in self.labels:
+            self.selected = self.labels[0]
+            raise ValueError(f"{event['new']} not one of {self.labels}")
+        self._clear_cache("selected_obj")
+        self._update_has_subregions()
+
+    def _update_has_subregions(self):
+        if "selected_has_subregions" in self._plugin_traitlets.keys():
+            if self.selected == "Entire Spectrum":
+                self.selected_has_subregions = False
+            else:
+                self.selected_has_subregions = len(self.selected_obj.subregions) > 1
+
+    @property
+    def labels(self):
+        return [s['label'] for s in self.items if 'label' in s.keys()]
+
+    @cached_property
+    def selected_obj(self):
+        if self.selected == "Entire Spectrum":
+            return None
+        return self.app.get_subsets_from_viewer("spectrum-viewer",
+                                                subset_type="spectral").get(self.selected)
+
+    def selected_min(self, cube):
+        if self.selected == "Entire Spectrum":
+            return np.nanmin(cube.spectral_axis.value)
+        else:
+            return self.selected_obj.lower.value
+
+    def selected_max(self, cube):
+        if self.selected == "Entire Spectrum":
+            return np.nanmax(cube.spectral_axis.value)
+        else:
+            return self.selected_obj.upper.value
```

Comment:
Will caching here really save anything? By the time this gets called, the whole cube is there. I think it is better to catch where you can prevent the cube from being recreated in the first place (most logically using its data label) but then you face the problem of cache invalidation. 🤷 

Also, do `selected_min` and `selected_max` need to be two different methods? Will there be a case where you only need one but not the other? Is it better to have a `selected_minmax` or `selected_cube_limits` (or whatever you want to call it) that returns both at once?

Reply:
In this case, it probably wouldn't be a significant savings to cache, unless the cube is really large.  But ease-of-caching is one thing we gain by this refactor, so I think we should use it when possible to keep ahead of optimization efforts.

I'm guessing this will get removed entirely if/when we support subsets with subregions in collapse/moment maps.  Since those are the only two cases that currently use it, I have no objection to switching to a single method.